### PR TITLE
8343535: IGV: Colorize nodes on demand

### PR DESCRIPTION
**Pull request by tobiasholenstein**
color

pick

nodes

Adds new option to IGV to color selected nodes:
1) select some nodes
2) `Ctrl + C` or `View` -> `Color action`
3) pick a color and apply

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Warning
&nbsp;⚠️ Patch contains a binary file (src/utils/IdealGraphVisualizer/View/src/main/resources/com/sun/hotspot/igv/view/images/color.gif)

### Issue
 * [JDK-8343535](https://bugs.openjdk.org/browse/JDK-8343535): IGV: Colorize nodes on demand (**Enhancement** - P4)

### Reviewers
 * [Christian Hagedorn](https://openjdk.org/census#chagedorn) (@chhagedorn - **Reviewer**)
 * [Roberto Castañeda Lozano](https://openjdk.org/census#rcastanedalo) (@robcasloz - **Reviewer**)

### Contributors
 * Roberto Castañeda Lozano `<rcastanedalo@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21925/head:pull/21925` \
`$ git checkout pull/21925`

Update a local copy of the PR: \
`$ git checkout pull/21925` \
`$ git pull https://git.openjdk.org/jdk.git pull/21925/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21925`

View PR using the GUI difftool: \
`$ git pr show -t 21925`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21925.diff">https://git.openjdk.org/jdk/pull/21925.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21925#issuecomment-2459977179)
</details>

### CODE REVIEW

**Review by chhagedorn — APPROVED** [review]
That's a nice feature! Works as expected on Linux with the short cut.

File: src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/DiagramScene.java [hunk 1]
```diff
@@ -222,6 +222,17 @@ public void filteredChanged(SelectionCoordinator coordinator) {
         }
     };
 
+    public void colorSelectedFigures(Color color) {
+        for (Figure figure :  model.getSelectedFigures()) {
```
Comment (turbanoff) [inline src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/DiagramScene.java:226]:
```suggestion
        for (Figure figure : model.getSelectedFigures()) {
```

**Review by robcasloz — CHANGES_REQUESTED** [review]
Nice improvement, thanks for working on this!

If the user selects a dark color, the node labels might become hard to read. Here's a simple change that addresses that by coloring the labels in white in that case. Please consider merging it into this PR:
```
diff --git a/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/widgets/FigureWidget.java b/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/widgets/FigureWidget.java
index a469d196a6b..495d844eb34 100644
--- a/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/widgets/FigureWidget.java
+++ b/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/widgets/FigureWidget.java
@@ -113,7 +113,6 @@ public FigureWidget(final Figure f, DiagramScene scene) {
             LayoutFactory.SerialAlignment.LEFT_TOP :
             LayoutFactory.SerialAlignment.CENTER;
         middleWidget.setLayout(LayoutFactory.createVerticalFlowLayout(textAlign, 0));
-        middleWidget.setBackground(f.getColor());
         middleWidget.setOpaque(true);
         middleWidget.getActions().addAction(new DoubleClickAction(this));
         middleWidget.setCheckClipping(false);
@@ -143,7 +142,6 @@ public FigureWidget(final Figure f, DiagramScene scene) {
             textWidget.addChild(lw);
             lw.setLabel(displayString);
             lw.setFont(Diagram.FONT);
-            lw.setForeground(getTextColor());
             lw.setAlignment(LabelWidget.Alignment.CENTER);
             lw.setVerticalAlignment(LabelWidget.VerticalAlignment.CENTER);
             lw.setBorder(BorderFactory.createEmptyBorder());
@@ -151,6 +149,8 @@ public FigureWidget(final Figure f, DiagramScene scene) {
         }
         formatExtraLabel(false);
 
+        refreshColor();
+
         if (getFigure().getWarning() != null) {
             ImageWidget warningWidget = new ImageWidget(scene, warningSign);
             Point warningLocation = new Point(getFigure().getWidth() - Figure.WARNING_WIDTH - Figure.INSET / 2, 0);
@@ -186,6 +186,9 @@ protected Sheet createSheet() {
 
     public void refreshColor() {
         middleWidget.setBackground(figure.getColor());
+        for (LabelWidget lw : labelWidgets) {
+            lw.setForeground(getTextColor());
+        }
     }
 
     @Override

```

**Comment by robcasloz** [comment]
In my opinion, the IGV toolbar is already pretty crowded (this hurts most when opening two graphs side-by-side) and I would prefer not adding the color icon there. On the other hand, we could add the action to the pop-up menu that's opened when right-clicking into a node or set of nodes. Here's my suggestion:

```
diff --git a/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/EditorTopComponent.java b/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/EditorTopComponent.java
index e68abd3297e..c4f2ac670e7 100644
--- a/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/EditorTopComponent.java
+++ b/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/EditorTopComponent.java
@@ -100,6 +100,7 @@ public EditorTopComponent(DiagramViewModel diagramViewModel) {
         };
 
         Action[] actionsWithSelection = new Action[]{
+                ColorAction.get(ColorAction.class),
                 ExtractAction.get(ExtractAction.class),
                 HideAction.get(HideAction.class),
                 null,
@@ -168,8 +169,6 @@ public void mouseMoved(MouseEvent e) {}
         toolBar.add(ReduceDiffAction.get(ReduceDiffAction.class));
         toolBar.add(ExpandDiffAction.get(ExpandDiffAction.class));
         toolBar.addSeparator();
-        toolBar.add(ColorAction.get(ColorAction.class));
-        toolBar.addSeparator();
         toolBar.add(ExtractAction.get(ExtractAction.class));
         toolBar.add(HideAction.get(HideAction.class));
         toolBar.add(ShowAllAction.get(ShowAllAction.class));
diff --git a/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ColorAction.java b/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ColorAction.java
index a51934a4322..92921c81512 100644
--- a/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ColorAction.java
+++ b/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ColorAction.java
@@ -43,7 +43,7 @@
         @ActionReference(path = "Shortcuts", name = "D-C")
 })
 @Messages({
-        "CTL_ColorAction=Color action",
+        "CTL_ColorAction=Color",
         "HINT_ColorAction=Color current set of selected nodes"
 })
 public final class ColorAction extends ModelAwareAction {
diff --git a/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ExtractAction.java b/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ExtractAction.java
index 24815527a0e..c2329cbb26f 100644
--- a/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ExtractAction.java
+++ b/src/utils/IdealGraphVisualizer/View/src/main/java/com/sun/hotspot/igv/view/actions/ExtractAction.java
@@ -42,7 +42,7 @@
         @ActionReference(path = "Shortcuts", name = "D-X")
 })
 @Messages({
-        "CTL_ExtractAction=Extract action",
+        "CTL_ExtractAction=Extract",
         "HINT_ExtractAction=Extract current set of selected nodes"
 })
 public final class ExtractAction extends ModelAwareAction {
```

@tobiasholenstein @chhagedorn what do you think? If you agree, feel free to merge the patch into this PR.

**Comment by chhagedorn** [comment]
> If the user selects a dark color, the node labels might become hard to read. Here's a simple change that addresses that by coloring the labels in white in that case. Please consider merging it into this PR:

Great idea!

> In my opinion, the IGV toolbar is already pretty crowded (this hurts most when opening two graphs side-by-side) and I would prefer not adding the color icon there

I agree with this. Maybe we should generally thinking about cleaning the toolbar in general and dropping some of the fewer used icons.
 
> On the other hand, we could add the action to the pop-up menu that's opened when right-clicking into a node or set of nodes. 

I thought about this, too. I think that would be quite handy and an intuitive thing to do when not being aware of the feature and checking if there is an option to do it.

**Comment by tobiasholenstein** [comment]
@robcasloz I have applied both your patches

**Review by robcasloz — APPROVED** [review]
Looks good, thanks Toby! I see that you also fixed the extra label color, nice! would be good to factor out that code together with that of `FigureWidget::getTextColor()`, but not a must.

**Review by chhagedorn — APPROVED** [review]
Looks good! Just tried it out. One more thing I've noticed: When selecting a color for a node and then trying to color another node, the color selection resets back to `#ffffff`. Would be nice if the last selection would have been stored. But I'm not sure how easy this is. Could also be done separately.

**Comment by tobiasholenstein** [comment]
> Looks good! Just tried it out. One more thing I've noticed: When selecting a color for a node and then trying to color another node, the color selection resets back to `#ffffff`. Would be nice if the last selection would have been stored. But I'm not sure how easy this is. Could also be done separately.

Screenshot 2024-11-08 at 15 27 06

right. I have updated it now to save the last colors and provide some default colors

**Comment by tobiasholenstein** [comment]
> Looks good, thanks Toby! I see that you also fixed the extra label color, nice! would be good to factor out that code together with that of `FigureWidget::getTextColor()`, but not a must.

done

**Comment by chhagedorn** [comment]
Great idea with the defaults! I've just tried it out on Linux. It does seem to remember my last choice and offers me some defaults. But somehow the window look off:

image

The defaults are hard to see and the rectangle saying "Preview" to the left is strange. It also says "Color Name: #ffffff" even though it chooses the last selected one correctly when pressing "OK".

**Review by chhagedorn — APPROVED** [review]
Now, after many tries, it seems to work! :-) Thanks for investigating further. Looks good!

**Review by robcasloz — APPROVED** [review]

**Comment by tobiasholenstein** [comment]
thanks for the reviews @robcasloz and @chhagedorn 

/integrate

**Pull request closed without merging**